lib/src/preprocess/polyline_preprocess_kernel.cu

```diff
@@ -234,8 +238,13 @@ cudaError_t polylinePreprocessWithTopkLauncher(
     B, L, P, outPointDim, tmpPolyline, tmpPolylineMask, tmpDistance);
 
   const dim3 blocks3(B, P, outPointDim);
-  const size_t sharedMemSize = sizeof(float) * L;
-  extractTopKPolylineKernel<<<blocks3, threadsPerBlock, sharedMemSize, stream>>>(
+  constexpr unsigned int itemsPerThread = 24;
+  if (threadsPerBlock * itemsPerThread < L) {
+    std::cerr << "Larger L (" << L << ") than acceptable range (< "
+              << threadsPerBlock * itemsPerThread << ") detected." << std::endl;
+    return cudaError_t::cudaErrorInvalidValue;
+  }
+  extractTopKPolylineKernel<threadsPerBlock, itemsPerThread><<<blocks3, threadsPerBlock>>>(
```

Comment:
[Q] This is just a question. Don't we have to put the kernel on the same stream with `kernel<<<...., stream>>>` in order to guarantee the order of function execution?

Reply:
I appreciate your detailed review!
Except for the default CUDA stream, we need to execute CUDA kernels in the same stream to guarantee the execution order. However, in terms of the default CUDA stream, if we do not generate a stream using [a special API](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1gb1e32aff9f59119e4d0a9858991c4ad3), the generated stream behavior will be synchronous with the default one.  
For example, 
```cuda
kernel_A<<<....., stream>>>(...);   // launch kernel_A in the stream
kernel_B<<<.....>>>(...);  // launch kernel_B in the default stream
kernel_C<<<....., stream>>>(...);  // launch kernel_C in the stream again
```
if we have this kind of code structure, `kernel_B` will wait to finish the execution of `kernel_A` implicitly.  In like manner, `kernel_C` execution will not start until `kernel_B` execution finishes.
For the reason of this implicit synchronization between CUDA streams, I believe the code you marked works as expected. (To be honest, I just missed to launch the kernel in the stream but fortunately it was valid code though)

BUT, since the code causes confusion, I modified the code to use the same stream as the  former/latter operations . Thank you very much for pointing this out!

Reply:
@manato Thank you for a clear explanation! I completely make sense :+1: